### ISSUE TITLE
It does not always work with the PHP compilation target

### ISSUE THREAD

**Issue by CedX** [benmerckx/watch#3]
No matter how hard I try, I can't get this library to work properly.

On CentOS 8, only the first build is triggered, the next builds are not triggered: I successfully get the "Build completed in Xs" message when launching the command (i.e. `haxe build.hxml -lib watch` or `lix watch build.hxml`), but if I modify a file, no further build is launched.

On Windows 10, the command either exits or fails:
- `haxe build.hxml -lib watch`: no "Build completed in Xs" message.
- `lix watch build.hxml`: results in `Failed to invoke 'haxelib run-dir watch [...]/haxe_libraries/watch/0.0.5/haxelib build.hxml' because Error: Command failed: haxelib run-dir watch [...]/haxe_libraries/watch/0.0.5/haxelib build.hxml`.

Any idea how to make it work?
My environment: Haxe 4.2.4 with `lix` 15.11.4 and `watch` 0.0.5.

**Comment by CedX** [comment]
Hmmm... with a dead simple test file, it works as expected on all tested platforms. So, there is something in my project dependencies that creates a conflict with this library...

**Comment by CedX** [comment]
I haven't found the exact source of this issue, but what I'm sure of that it's related to some macros run by one or several of my dependencies (mostly Tinkerbell ones), and the order of their execution. I have strictly no idea how to debug that :cry:

**Comment by benmerckx** [comment]
Could you add all used tink libraries to your dead simple test file and see if you can reproduce that way? It probably is some macro interfering with the watch macro but hard to debug without reproduction :)

**Comment by CedX** [comment]
Of course! I will try to create a project for you that shows the problem. My "dead simple test file" is currently far too simple for reproducing this issue (i.e. I used: `function main() trace("OK");`).

A preview of the dependencies I currently use (...isolating the problematic one is hard):
![image](https://user-images.githubusercontent.com/1643950/143059848-3e3ac458-c22b-4454-9eab-172b0db548cf.png)

**Comment by CedX** [comment]
Here is a first try at isolating the root cause:
https://github.com/cedx/watch-issue-3

I've added `tink_web` as first dependency (i.e. `lix +lib tink_web`)... and well, there's quite a few transitive dependencies...
The issue is already there (i.e. on Windows : `haxe -lib watch build.hxml` exits immediately, and `lix watch build.hxml` results in an error).

**Comment by CedX** [comment]
Interesting: if I use a Node.js container (i.e. the compilation target is JavaScript instead of PHP), `watch` works as expected.
Cf. https://github.com/cedx/watch-issue-3/tree/nodejs

And now I realize that the only projects where `watch` failed were all using PHP as compilation target.

**Comment by CedX** [comment]
I give up: there is something wrong in the Haxe compilation/completion server when targetting PHP!
This works:

```haxe
package project;

import tink.Json;

typedef Sample = {
	var prop: String;
}

class Main {
	static function main() {
		var foo: Sample = Json.parse('{"prop": "foo"');
		trace(foo);
	}
}
```

But not that (i.e. SAME code but with comments):

```haxe
package project;

// import tink.http.Response.OutgoingResponse;
// import tink.http.containers.PhpContainer;
// import tink.web.routing.Context;
// import tink.web.routing.Router;

// import tink.QueryString;
import tink.Json;
//import tink.Url;

typedef Sample = {
	var prop: String;
}

class Main {
	static function main() {
		var foo: Sample = Json.parse('{"prop": "foo"');
		trace(foo);

		//var bar: Url = "https://codeurs.be";
		//trace(bar);

		//trace(QueryString.build({foo: "bar"}));

		/*
		final router = new Router<Root>(new Root());
		PhpContainer.inst.run(request -> router.route(Context.ofRequest(request)).recover(error -> {
			if (error.code == InternalError) trace(error);
			OutgoingResponse.reportError(error);
		}));*/
	}
}
```

If I remove the line `//import tink.Url;`, everything works. If I restore this line, I get a compilation error. If I relaunch the `lix watch` command, everything works.

**That's completely random and insane!**

Note: I've also a JS project that compiles fine, but when using VSCode or `watch`, I always get errors like `Type name haxe.PosInfos is redefined from module haxe.PosInfos` or `Type name tink.Annex is redefined from module tink.CoreAp`. That's why I think the problem lies more in the compilation/completion server than in the `watch` library (cf. https://haxe.org/manual/cr-completion-server.html: the compilation server can have some side effects on macro execution).

**Comment by benmerckx** [comment]
This took a long time but I've found the root cause here. To react to changes we watch every class path. This includes the root dir because you can always import directly from the root (`import X -> ./X.hx`). However the target generated files will often also reside in the root dir (ex. `/dist` or in your case even straight into root). What happens in the case of php compilation is the way Haxe generates the php output files somehow crashes the file watcher in a way we cannot recover from (the compile process crashes without warning/error). I'm using the very new [FsEvent](https://api.haxe.org/eval/luv/FsEvent.html) to watch for file changes. The core issue might be fixed in future versions (but it'd be hard to file a good bug report here, as it might also be an issue with the luv library). Possible workarounds:
- Somehow not watch the target output directories. Unfortunately there's no exclude option in the file watcher and in addition (like in your example) they could fully overlap.
- Use a different file watcher, but performance might suffer (there's this for example: https://github.com/vegardit/haxe-files#file-watching)
- Temporarily stop file watching during compilation, however then we can't respond to file changes while compiling (which can happen often)

Not sure how to fix yet :)

**Comment by benmerckx** [comment]
I've added a flag to exclude watching the root which you can use like this:
````
haxe build.hxml -lib watch -D watch.excludeRoot
````
This should make the lib useable for your use case until the root cause gets fixed.

**Comment by CedX** [comment]
Wonderful @benmerckx, the latest `watch` version fixed my issue with the current project (a pure PHP one). I just have to test with other more complex projects (JS + PHP, that is 2 simultaneous `watch` tasks...).
Thanks for your efforts!

**Comment by benmerckx** [comment]
Additionally, I suspect there might be an issue with watching subdirectories on linux as [it seems](http://docs.libuv.org/en/v1.x/fs_event.html#c.uv_fs_event_start) the recursive flag will only work on MacOS & Windows. Which would explain only the initial build working on CentOS 8.

**Comment by CedX** [comment]
Ah yes: I've only tested on Windows. Effectively on CentOS 8, the initial build is still the only one triggered.
Same issue with the Node.js file watcher, also based on `libuv` :
https://nodejs.org/api/fs.html#fswatchfilename-options-listener (see "Caveats")

**Comment by benmerckx** [comment]
This should be okay in 0.1.0 now.
